### ISSUE TITLE
False-positive while_immutable_condition lint

### ISSUE THREAD

**Issue by IsaacWoods** [rust-lang-nursery/rust-clippy#2584]
I have the following code:
``` rust
let ptr : *const u8 = ...;
let mut len = 0;
while unsafe { *ptr.offset(len) } != 0
{
    len += 1;
}
```

Clearly, `len` is incremented; however, I get this lint:
```
warning: Variable in the condition are not mutated in the loop body. This either leads to an infinite
 or to a never running loop.
  --> src/multiboot2/elf_sections.rs:58:32
   |
58 |                          while unsafe { *ptr.offset(len) } != 0
   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: #[warn(while_immutable_condition)] on by default
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.190/ind
ex.html#while_immutable_condition
```

This is with `clippy v0.0.190` running on Rust nightly (`rustc 1.26.0-nightly (9c9424de5 2018-03-27)`)

**Comment by oli-obk** [comment]
we should just bail out on raw pointer derefs
